### PR TITLE
For #480: Created soldier counter attack card

### LINKED ISSUE #480: Solider Card - Implement Counter-Attack
Counter-Attack - 1 Ally - During the turn, whenever target is attacked, fire back for 1 Ammo, 2x ATK

Size: 2 units

### PR DESCRIPTION
**Pull request by paulodamaso**
Closes #480:
- Created soldier counter attack card
- Created  `Fireback` effect
- Created tests

Couldn't understand resource logic, so test about resource is commented.

This PR depends on #391 resolution

### CODE REVIEW

**Comment by paulodamaso** [comment]
@SilasReinagel Fixed the design using the effect data evolution introduced in #486

File: src/ProjectNeon/Assets/Features/Effects/ConditionalEffect.cs [hunk 1]
```diff
@@ -0,0 +1,28 @@
+﻿using System;
+using UnityEngine;
+
+public abstract class ConditionalEffect : Effect
+{
+    protected Effect _effect;
+    protected Member _source;
```
Comment (SilasReinagel) [inline src/ProjectNeon/Assets/Features/Effects/ConditionalEffect.cs]: Member and Target should not be stored as fields.

File: src/ProjectNeon/Assets/Features/Effects/CostPrimaryResourceEffect.cs [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+    public CostPrimaryResourceEffect(Effect effect, int cost)
+    {
+        _effect = effect;
+        _cost = cost;
+    }
+
+    void Effect.Apply(Member source, Target target)
+    {
+        _effect.Apply(source, target);
+        source.State.SpendPrimaryResource(_cost);
```
Comment (SilasReinagel) [inline src/ProjectNeon/Assets/Features/Effects/CostPrimaryResourceEffect.cs]: How do we ensure that we don't perform this action if the cost can't be paid?

File: src/ProjectNeon/Assets/Features/Effects/WithCostOfBasicResource.cs [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+
+    public WithCostOfBasicResource (Effect effect, int quantity)
+    {
+        _effect = effect;
+        _quantity = quantity;
+    }
+
+    void Effect.Apply(Member source, Target target)
+    {
+        throw new System.NotImplementedException();
```
Comment (SilasReinagel) [inline src/ProjectNeon/Assets/Features/Effects/WithCostOfBasicResource.cs]: This is not implemented.

File: src/ProjectNeon/Assets/Features/Effects/CostPrimaryResourceEffect.cs (hunk 2 above)
Comment (paulodamaso) [reply src/ProjectNeon/Assets/Features/Effects/CostPrimaryResourceEffect.cs]: @SilasReinagel I think that it shouldn't be made here. The UI should't even let us pick a card if we didn't have the enough resources to play it. If we instantiated this cost effect then we already have the sufficient resource to play it

File: src/ProjectNeon/Assets/Features/Effects/WithCostOfBasicResource.cs (hunk 3 above)
Comment (paulodamaso) [reply src/ProjectNeon/Assets/Features/Effects/WithCostOfBasicResource.cs]: @SilasReinagel Sorry this class was a development prototype that I forgot do remove

File: src/ProjectNeon/Assets/Features/Effects/CostPrimaryResourceEffect.cs (hunk 2 above)
Comment (SilasReinagel) [reply src/ProjectNeon/Assets/Features/Effects/CostPrimaryResourceEffect.cs]: @paulodamaso This is the cost of a triggered effect, not the cost of the card. The card has already been played and paid for.

Nothing can predict how many time Counter-Attack will trigger. The Solider is allowed to Counter each attack, until he runs out of ammo. Then, obviously, he isn't allowed to keep counter-attacking with his empty gun...

File: src/ProjectNeon/Assets/Features/Effects/ConditionalEffect.cs (hunk 1 above)
Comment (SilasReinagel) [reply src/ProjectNeon/Assets/Features/Effects/ConditionalEffect.cs]: This is still open.

**Comment by paulodamaso** [comment]
@SilasReinagel Left a test implemented on how the creation of a primary resource with a starting value is not being executed correctly.

**Pull request merged**